# Data Wrangling 

As I complete the module about data wrangling, now I have to do some exercise to let my knowledge from the module flow

## Gathering Data

The first think I have to do is get the data. The data provides by Dicoding with 4 kinds of dataset. Here's the [link](https://github.com/dicodingacademy/dicoding_dataset/tree/main/DicodingCollection) of the datasets.

So now I have the dataset and ready to use in this notebook.

In [1]:
import pandas as pd

custumer_df = pd.read_csv("../Datasets/customers.csv")
custumer_df

customer_id customer_name             gender  age  \
0               1       fulan 1             Female   30   
1               2       fulan 2  Prefer not to say   69   
2               3       fulan 3  Prefer not to say   59   
3               4       fulan 4  Prefer not to say   67   
4               5       fulan 5  Prefer not to say   30   
...           ...           ...                ...  ...   
1002          996     fulan 996  Prefer not to say   59   
1003          997     fulan 997  Prefer not to say   30   
1004          998     fulan 998  Prefer not to say   32   
1005          999     fulan 999  Prefer not to say   30   
1006         1000    fulan 1000               Male   71   

                        home_address  zip_code                city  \
0     8606 Victoria TerraceSuite 560      5464       Johnstonhaven   
1         8327 Kirlin SummitApt. 461      8223     New Zacharyfort   
2          269 Gemma SummitSuite 109      5661            Aliburgh   
3          743 Bailey GroveSuite 141      1729   South Justinhaven   
4            48 Hyatt ManorSuite 375      4032      Griffithsshire   
...                              ...       ...                 ...   
1002    0433 Armstrong HillSuite 974      7613     Lake Danielland   
1003         04 Howell PassSuite 209      6950         Ellaborough   
1004       72 Annabelle PassApt. 446        52          Kohlerberg   
1005       170 Wilson AvenueApt. 577      7849      East Oscarfurt   
1006      1671 Lauren KnollSuite 945      9012  Lake Audreyborough   

                             state    country  
0               Northern Territory  Australia  
1                  South Australia  Australia  
2     Australian Capital Territory  Australia  
3                       Queensland  Australia  
4                       Queensland  Australia  
...                            ...        ...  
1002                      Tasmania  Australia  
1003                      Tasmania  Australia  
1004                    Queensland  Australia  
1005             Western Australia  Australia  
1006                      Tasmania  Australia  

[1007 rows x 9 columns]

In [10]:
orders_df = pd.read_csv("../Datasets/orders.csv")
orders_df

order_id  customer_id  payment order_date delivery_date
0           1           64    30811  2021-8-30    2021-09-24
1           2          473    50490   2021-2-3    2021-02-13
2           3          774    46763  2021-10-8    2021-11-03
3           4          433    39782   2021-5-6    2021-05-19
4           5          441    14719  2021-3-23    2021-03-24
..        ...          ...      ...        ...           ...
995       996          345    37843  2021-1-13    2021-02-02
996       997          346    53831  2021-1-18    2021-01-31
997       998          407    53308   2021-5-5    2021-05-21
998       999          428    31643  2021-6-15    2021-07-12
999      1000          896    27836   2021-4-7    2021-04-24

[1000 rows x 5 columns]

In [11]:
product_df = pd.read_csv("../Datasets/products.csv")
product_df

product_id product_type       product_name size  colour  price  \
0              0        Shirt       Oxford Cloth   XS     red    114   
1              1        Shirt       Oxford Cloth    S     red    114   
2              2        Shirt       Oxford Cloth    M     red    114   
3              3        Shirt       Oxford Cloth    L     red    114   
4              4        Shirt       Oxford Cloth   XL     red    114   
...          ...          ...                ...  ...     ...    ...   
1261        1255     Trousers  Tracksuit Bottoms   XS  violet     91   
1262        1256     Trousers  Tracksuit Bottoms    S  violet     91   
1263        1257     Trousers  Tracksuit Bottoms    M  violet     91   
1264        1258     Trousers  Tracksuit Bottoms    L  violet     91   
1265        1259     Trousers  Tracksuit Bottoms   XL  violet     91   

      quantity                                        description  
0           66       A red coloured, XS sized, Oxford Cloth Shirt  
1           53        A red coloured, S sized, Oxford Cloth Shirt  
2           54        A red coloured, M sized, Oxford Cloth Shirt  
3           69        A red coloured, L sized, Oxford Cloth Shirt  
4           47       A red coloured, XL sized, Oxford Cloth Shirt  
...        ...                                                ...  
1261        67  A violet coloured, XS sized, Tracksuit Bottoms...  
1262        48  A violet coloured, S sized, Tracksuit Bottoms ...  
1263        73  A violet coloured, M sized, Tracksuit Bottoms ...  
1264        45  A violet coloured, L sized, Tracksuit Bottoms ...  
1265        60  A violet coloured, XL sized, Tracksuit Bottoms...  

[1266 rows x 8 columns]

In [12]:
sales_df = pd.read_csv("../Datasets/sales.csv")
sales_df

sales_id  order_id  product_id  price_per_unit  quantity  total_price
0            0         1         218             106         2        212.0
1            1         1         481             118         1        118.0
2            2         1           2              96         3        288.0
3            3         1        1002             106         2        212.0
4            4         1         691             113         3        339.0
...        ...       ...         ...             ...       ...          ...
4995      4995       998         321             109         2        218.0
4996      4996       998         251              95         3        285.0
4997      4997       999         872             113         1        113.0
4998      4998       999         998             106         2        212.0
4999      4999       999        1105             115         1        115.0

[5000 rows x 6 columns]

Okay, looks like the dataset related to each other and bound by id's

It's time for doing assessing data on this datasets

## Assessing Data

### Checking Customer Dataframe

In [13]:
custumer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customer_id    1007 non-null   int64 
 1   customer_name  1007 non-null   object
 2   gender         989 non-null    object
 3   age            1007 non-null   int64 
 4   home_address   1007 non-null   object
 5   zip_code       1007 non-null   int64 
 6   city           1007 non-null   object
 7   state          1007 non-null   object
 8   country        1007 non-null   object
dtypes: int64(3), object(6)
memory usage: 70.9+ KB


Customer dataframe had some missing value on it

In [14]:
custumer_df.isna().sum()

customer_id       0
customer_name     0
gender           18
age               0
home_address      0
zip_code          0
city              0
state             0
country           0
dtype: int64

there are 18 missing value on gender column

In [23]:
print("Duplicated value: " ,custumer_df.duplicated().sum())

Duplicated value:  6


In [18]:
custumer_df.describe()

customer_id          age     zip_code
count  1007.000000  1007.000000  1007.000000
mean    501.726912    50.929494  5012.538232
std     288.673238    30.516299  2885.836112
min       1.000000    20.000000     2.000000
25%     252.500000    34.000000  2403.500000
50%     502.000000    50.000000  5087.000000
75%     751.500000    65.000000  7493.500000
max    1000.000000   700.000000  9998.000000

there is something wrong with the age, The max value of that colomn is 700, look's like there are some inacurate value on that column

### Checking Order Dataframe

In [19]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   order_id       1000 non-null   int64 
 1   customer_id    1000 non-null   int64 
 2   payment        1000 non-null   int64 
 3   order_date     1000 non-null   object
 4   delivery_date  1000 non-null   object
dtypes: int64(3), object(2)
memory usage: 39.2+ KB


It looks like there's no missing value on this dataframe but There are something wrong with Dtype of dates, it should be in date Dtype not object 

In [22]:
print("Duplicated value: ", orders_df.duplicated().sum())

Duplicated value:  0


In [24]:
orders_df.describe()

order_id  customer_id       payment
count  1000.000000  1000.000000   1000.000000
mean    500.500000   506.640000  33972.936000
std     288.819436   277.115502  14451.609047
min       1.000000     1.000000  10043.000000
25%     250.750000   275.250000  21329.250000
50%     500.500000   515.000000  33697.500000
75%     750.250000   737.250000  46249.000000
max    1000.000000  1000.000000  59910.000000

Okay, I think there's nothing wrong with this dataframe, except the Dtype of dates

### Checking Product Dataframe

In [25]:
product_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1266 entries, 0 to 1265
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_id    1266 non-null   int64 
 1   product_type  1266 non-null   object
 2   product_name  1266 non-null   object
 3   size          1266 non-null   object
 4   colour        1266 non-null   object
 5   price         1266 non-null   int64 
 6   quantity      1266 non-null   int64 
 7   description   1266 non-null   object
dtypes: int64(3), object(5)
memory usage: 79.2+ KB


No problem here

In [27]:
print("Duplicated value: ", product_df.duplicated().sum())

Duplicated value:  6


In [28]:
product_df.describe()

product_id        price     quantity
count  1266.000000  1266.000000  1266.000000
mean    627.926540   105.812006    60.138231
std     363.971586     9.715611    11.682791
min       0.000000    90.000000    40.000000
25%     313.250000    95.250000    50.000000
50%     626.500000   109.000000    60.000000
75%     942.750000   114.000000    70.000000
max    1259.000000   119.000000    80.000000

Okay I think duplicated value is the only problem in this dataframe

### Checking Sales Dataframe

In [29]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sales_id        5000 non-null   int64  
 1   order_id        5000 non-null   int64  
 2   product_id      5000 non-null   int64  
 3   price_per_unit  5000 non-null   int64  
 4   quantity        5000 non-null   int64  
 5   total_price     4981 non-null   float64
dtypes: float64(1), int64(5)
memory usage: 234.5 KB


In [30]:
sales_df.isna().sum()

sales_id           0
order_id           0
product_id         0
price_per_unit     0
quantity           0
total_price       19
dtype: int64

There are missing value on total_price

In [31]:
print("Duplicated value: ", sales_df.duplicated().sum())

Duplicated value:  0


In [32]:
sales_df.describe()

sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  4981.000000  
mean    206.307368  
std      86.352449  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

Okay just missing value on this dataframe

## Cleaning Data

### Cleaning Customer Dataframe

First let see if the missing data is in object or in data

In [34]:
custumer_df.gender.info()

<class 'pandas.core.series.Series'>
RangeIndex: 1007 entries, 0 to 1006
Series name: gender
Non-Null Count  Dtype 
--------------  ----- 
989 non-null    object
dtypes: object(1)
memory usage: 8.0+ KB


okay it's an object data, so we have to replace the missing value with the most frequent data 

In [35]:
custumer_df.gender.value_counts()

gender
Prefer not to say    731
Male                 143
Female               115
Name: count, dtype: int64

and if we see what's the unique value of this colomn it's a "Prefer not to say" so the missing value will be replace with it

for this problem we use 2 method, with pandas or with scikit-learn, I prefer scikit-learn, but in this problem i'm gonna use pandas first

In [36]:
custumer_df.fillna(value="Prefer not to say", inplace=True)

In [37]:
custumer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customer_id    1007 non-null   int64 
 1   customer_name  1007 non-null   object
 2   gender         1007 non-null   object
 3   age            1007 non-null   int64 
 4   home_address   1007 non-null   object
 5   zip_code       1007 non-null   int64 
 6   city           1007 non-null   object
 7   state          1007 non-null   object
 8   country        1007 non-null   object
dtypes: int64(3), object(6)
memory usage: 70.9+ KB


okay looks like there are no missing value anymore

Next, we gonna drop the duplicated value

In [46]:
custumer_df.drop_duplicates(inplace=True)

In [48]:
custumer_df.duplicated().sum()

0

Duplicated value solved

Next, we gonna solve inacurate value on age

In [38]:
custumer_df[custumer_df.age == custumer_df.age.max()]

customer_id customer_name             gender  age  \
967          961     fulan 961  Prefer not to say  700   

                   home_address  zip_code        city            state  \
967  29 Farrell ParadeSuite 818      6528  New Joseph  South Australia   

       country  
967  Australia

here's the inacurate value column, maybe it use 0 to much, let's change it to 70

In [39]:
custumer_df.age.replace(custumer_df.age.max(), 70, inplace=True)

In [40]:
custumer_df.describe()

customer_id          age     zip_code
count  1007.000000  1007.000000  1007.000000
mean    501.726912    50.303873  5012.538232
std     288.673238    22.637014  2885.836112
min       1.000000    20.000000     2.000000
25%     252.500000    34.000000  2403.500000
50%     502.000000    50.000000  5087.000000
75%     751.500000    65.000000  7493.500000
max    1000.000000   500.000000  9998.000000

There's still inacurate value, let's change it

In [41]:
custumer_df[custumer_df.age == custumer_df.age.max()]

customer_id customer_name             gender  age  \
215          216     fulan 216  Prefer not to say  500   

               home_address  zip_code       city               state  \
215  038 Haley MewsApt. 810      3991  Bayertown  Northern Territory   

       country  
215  Australia

In [42]:
custumer_df.age.replace(custumer_df.age.max(), 50, inplace=True)

In [43]:
custumer_df.describe()

customer_id          age     zip_code
count  1007.000000  1007.000000  1007.000000
mean    501.726912    49.857001  5012.538232
std     288.673238    17.641262  2885.836112
min       1.000000    20.000000     2.000000
25%     252.500000    34.000000  2403.500000
50%     502.000000    50.000000  5087.000000
75%     751.500000    64.500000  7493.500000
max    1000.000000    80.000000  9998.000000

okay inacurate value solved

### Cleaning Order Dataframe

The problem of this dataframe is in Dtype, where dates should in date Dtype not object

Let's change the Dtype

In [44]:
miss_dtype_column = ["order_date", "delivery_date"]

for column in miss_dtype_column:
    orders_df[column] = pd.to_datetime(orders_df[column])

In [45]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   order_id       1000 non-null   int64         
 1   customer_id    1000 non-null   int64         
 2   payment        1000 non-null   int64         
 3   order_date     1000 non-null   datetime64[ns]
 4   delivery_date  1000 non-null   datetime64[ns]
dtypes: datetime64[ns](2), int64(3)
memory usage: 39.2 KB


Okay solved

### Cleaning Product Dataframe

The only problem in this dataframe is duplicated value, so let's solve it

In [49]:
product_df.drop_duplicates(inplace=True)

In [50]:
product_df.duplicated().sum()

0

Okay solved

### Cleaning Sales Dataframe

Okay the only problem in this dataframe is missing value, let's solve it

In [51]:
sales_df[sales_df.total_price.isna()]

sales_id  order_id  product_id  price_per_unit  quantity  total_price
9            9         2        1196             105         1          NaN
121        121        27        1027              90         3          NaN
278        278        63         360              94         2          NaN
421        421        95        1091             115         1          NaN
489        489       108        1193             105         3          NaN
539        539       117         405             119         2          NaN
636        636       134         653              93         3          NaN
687        687       145        1138             102         1          NaN
854        854       177          64             104         1          NaN
1079      1079       222         908              94         3          NaN
1193      1193       248        1121             102         2          NaN
1313      1313       272         826             117         1          NaN
1548      1548       316         103             118         3          NaN
1688      1688       345         428             107         1          NaN
1775      1775       359         694             113         2          NaN
1902      1902       381        1218             105         3          NaN
2025      2025       408         611             112         3          NaN
2164      2164       436         583             100         3          NaN
2347      2347       476         696             113         2          NaN

Okay as we know that total price is multiplication between price_pre_unit and quantity, so let's fill it 

In [52]:
sales_df["total_price"] = sales_df["price_per_unit"] * sales_df["quantity"]

In [54]:
sales_df.isna().sum()

sales_id          0
order_id          0
product_id        0
price_per_unit    0
quantity          0
total_price       0
dtype: int64

Okay it's all clean

# Eksplolatory Data Analysis

So we have the data clean, it's time to explore the data by giving a question on in. By knowing the question we will know what's the target or what's the output we gonna make from this dataset

Question:
1. How's the performance of the sales and revenue of the corporate a few months ago?
2. What product sold the most and the least?
3. How's the demography of our customer?
4. When was the last transaction of our customer?
5. How often a customer boughts a few months ago?
6. How much money customer spend a few months ago?

## Explorate Data

### Exploring Customer Dataframe

this dataframe contains our customer information such as age, gender, and location

In [56]:
custumer_df.describe(include='all')

customer_id customer_name             gender          age  \
count   1000.000000          1000               1000  1000.000000   
unique          NaN          1000                  3          NaN   
top             NaN       fulan 1  Prefer not to say          NaN   
freq            NaN             1                742          NaN   
mean     500.500000           NaN                NaN    49.860000   
std      288.819436           NaN                NaN    17.647828   
min        1.000000           NaN                NaN    20.000000   
25%      250.750000           NaN                NaN    34.000000   
50%      500.500000           NaN                NaN    50.000000   
75%      750.250000           NaN                NaN    65.000000   
max     1000.000000           NaN                NaN    80.000000   

                          home_address     zip_code         city  \
count                             1000  1000.000000         1000   
unique                            1000          NaN          961   
top     8606 Victoria TerraceSuite 560          NaN  East Sophia   
freq                                 1          NaN            3   
mean                               NaN  5004.872000          NaN   
std                                NaN  2884.497332          NaN   
min                                NaN     2.000000          NaN   
25%                                NaN  2401.750000          NaN   
50%                                NaN  5083.000000          NaN   
75%                                NaN  7460.250000          NaN   
max                                NaN  9998.000000          NaN   

                  state    country  
count              1000       1000  
unique                8          1  
top     South Australia  Australia  
freq                139       1000  
mean                NaN        NaN  
std                 NaN        NaN  
min                 NaN        NaN  
25%                 NaN        NaN  
50%                 NaN        NaN  
75%                 NaN        NaN  
max                 NaN        NaN

From this table, we know that our customer's age is from 20's to 80's with mean 49,86 and std 17,64. That's very large segments of market

now we gonna make a pivot table to know a gender and the age of our customer

In [57]:
custumer_df.groupby(by='gender').agg({
    "customer_id": "nunique",
    "age": ["min", "max", "mean", "std"]
})

customer_id age                          
                      nunique min max       mean        std
gender                                                     
Female                    115  20  79  49.147826  16.646607
Male                      143  20  80  51.230769  18.462635
Prefer not to say         742  20  80  49.706199  17.648386

Almost same as the table before

Next, we gonna try to see distribution of our customer's location by pivot table

In [58]:
custumer_df.groupby(by="city").customer_id.nunique().sort_values(ascending=False)

city
East Aidan       3
East Sophia      3
New Ava          3
West Jackfort    2
Corkeryshire     2
                ..
Jordanside       1
Josephland       1
Josephmouth      1
Justinport       1
Zacville         1
Name: customer_id, Length: 961, dtype: int64

In [59]:
custumer_df.groupby(by="state").customer_id.nunique().sort_values(ascending=False)

state
South Australia                 139
Queensland                      134
New South Wales                 132
Northern Territory              125
Western Australia               124
Australian Capital Territory    121
Victoria                        121
Tasmania                        104
Name: customer_id, dtype: int64

As we see that our customer much located at East Aidan, East Sophia, and New Ava. For the country, most of them located at South Australia, Queensland, and New South Wales 

### Exploring Sales Dataframe

This dataframe contain information about The orders id, The orders date, and the orders delivery

First, I want to know how much time do we need to travel our product by counting the delivery time

In [60]:
delivery_time = orders_df["delivery_date"] - orders_df["order_date"]

In [61]:
delivery_time

0     25 days
1     10 days
2     26 days
3     13 days
4      1 days
        ...  
995   20 days
996   13 days
997   16 days
998   27 days
999   17 days
Length: 1000, dtype: timedelta64[ns]

In [63]:
orders_df["delivery_time"] = delivery_time
orders_df

order_id  customer_id  payment order_date delivery_date delivery_time
0           1           64    30811 2021-08-30    2021-09-24       25 days
1           2          473    50490 2021-02-03    2021-02-13       10 days
2           3          774    46763 2021-10-08    2021-11-03       26 days
3           4          433    39782 2021-05-06    2021-05-19       13 days
4           5          441    14719 2021-03-23    2021-03-24        1 days
..        ...          ...      ...        ...           ...           ...
995       996          345    37843 2021-01-13    2021-02-02       20 days
996       997          346    53831 2021-01-18    2021-01-31       13 days
997       998          407    53308 2021-05-05    2021-05-21       16 days
998       999          428    31643 2021-06-15    2021-07-12       27 days
999      1000          896    27836 2021-04-07    2021-04-24       17 days

[1000 rows x 6 columns]

In [64]:
orders_df.describe(include='all')

order_id  customer_id       payment                  order_date  \
count  1000.000000  1000.000000   1000.000000                        1000   
mean    500.500000   506.640000  33972.936000  2021-05-27 18:38:52.800000   
min       1.000000     1.000000  10043.000000         2021-01-01 00:00:00   
25%     250.750000   275.250000  21329.250000         2021-03-13 18:00:00   
50%     500.500000   515.000000  33697.500000         2021-05-27 12:00:00   
75%     750.250000   737.250000  46249.000000         2021-08-12 06:00:00   
max    1000.000000  1000.000000  59910.000000         2021-10-24 00:00:00   
std     288.819436   277.115502  14451.609047                         NaN   

             delivery_date              delivery_time  
count                 1000                       1000  
mean   2021-06-10 20:31:12    14 days 01:52:19.200000  
min    2021-01-03 00:00:00            1 days 00:00:00  
25%    2021-03-28 00:00:00            8 days 00:00:00  
50%    2021-06-11 12:00:00           14 days 00:00:00  
75%    2021-08-24 06:00:00           21 days 00:00:00  
max    2021-11-20 00:00:00           27 days 00:00:00  
std                    NaN  7 days 16:58:24.240654075

Look's like delivery time is very confusing to analyze, let's transform it so that it easy to analyze

In [70]:
delivery_time = orders_df["delivery_date"] - orders_df["order_date"]
delivery_time = delivery_time.apply(lambda x: x.total_seconds())
orders_df['delivery_time'] = round(delivery_time/86400)
orders_df

order_id  customer_id  payment order_date delivery_date  delivery_time
0           1           64    30811 2021-08-30    2021-09-24           25.0
1           2          473    50490 2021-02-03    2021-02-13           10.0
2           3          774    46763 2021-10-08    2021-11-03           26.0
3           4          433    39782 2021-05-06    2021-05-19           13.0
4           5          441    14719 2021-03-23    2021-03-24            1.0
..        ...          ...      ...        ...           ...            ...
995       996          345    37843 2021-01-13    2021-02-02           20.0
996       997          346    53831 2021-01-18    2021-01-31           13.0
997       998          407    53308 2021-05-05    2021-05-21           16.0
998       999          428    31643 2021-06-15    2021-07-12           27.0
999      1000          896    27836 2021-04-07    2021-04-24           17.0

[1000 rows x 6 columns]

In [71]:
orders_df.describe()

order_id  customer_id       payment                  order_date  \
count  1000.000000  1000.000000   1000.000000                        1000   
mean    500.500000   506.640000  33972.936000  2021-05-27 18:38:52.800000   
min       1.000000     1.000000  10043.000000         2021-01-01 00:00:00   
25%     250.750000   275.250000  21329.250000         2021-03-13 18:00:00   
50%     500.500000   515.000000  33697.500000         2021-05-27 12:00:00   
75%     750.250000   737.250000  46249.000000         2021-08-12 06:00:00   
max    1000.000000  1000.000000  59910.000000         2021-10-24 00:00:00   
std     288.819436   277.115502  14451.609047                         NaN   

             delivery_date  delivery_time  
count                 1000    1000.000000  
mean   2021-06-10 20:31:12      14.078000  
min    2021-01-03 00:00:00       1.000000  
25%    2021-03-28 00:00:00       8.000000  
50%    2021-06-11 12:00:00      14.000000  
75%    2021-08-24 06:00:00      21.000000  
max    2021-11-20 00:00:00      27.000000  
std                    NaN       7.707225

Okay now we know that our delivery time ranged from 1 day to 27 days 

### Exploring Order and Customer Data

As we see in order dataframe there's an order_id and customer_id, it means that this customer once buy our product, Let's mark which customer that once buy our product with adding column status. if it's once buy our product the status is "Active" if not it's "Non active"

In [72]:
customer_buy = orders_df.customer_id.tolist()
custumer_df["status"] = custumer_df["customer_id"].apply(lambda x: "Active" if x in customer_buy else "Non Active")
custumer_df

customer_id customer_name             gender  age  \
0               1       fulan 1             Female   30   
1               2       fulan 2  Prefer not to say   69   
2               3       fulan 3  Prefer not to say   59   
3               4       fulan 4  Prefer not to say   67   
4               5       fulan 5  Prefer not to say   30   
...           ...           ...                ...  ...   
1002          996     fulan 996  Prefer not to say   59   
1003          997     fulan 997  Prefer not to say   30   
1004          998     fulan 998  Prefer not to say   32   
1005          999     fulan 999  Prefer not to say   30   
1006         1000    fulan 1000               Male   71   

                        home_address  zip_code                city  \
0     8606 Victoria TerraceSuite 560      5464       Johnstonhaven   
1         8327 Kirlin SummitApt. 461      8223     New Zacharyfort   
2          269 Gemma SummitSuite 109      5661            Aliburgh   
3          743 Bailey GroveSuite 141      1729   South Justinhaven   
4            48 Hyatt ManorSuite 375      4032      Griffithsshire   
...                              ...       ...                 ...   
1002    0433 Armstrong HillSuite 974      7613     Lake Danielland   
1003         04 Howell PassSuite 209      6950         Ellaborough   
1004       72 Annabelle PassApt. 446        52          Kohlerberg   
1005       170 Wilson AvenueApt. 577      7849      East Oscarfurt   
1006      1671 Lauren KnollSuite 945      9012  Lake Audreyborough   

                             state    country      status  
0               Northern Territory  Australia      Active  
1                  South Australia  Australia  Non Active  
2     Australian Capital Territory  Australia  Non Active  
3                       Queensland  Australia  Non Active  
4                       Queensland  Australia  Non Active  
...                            ...        ...         ...  
1002                      Tasmania  Australia      Active  
1003                      Tasmania  Australia  Non Active  
1004                    Queensland  Australia      Active  
1005             Western Australia  Australia  Non Active  
1006                      Tasmania  Australia      Active  

[1000 rows x 10 columns]

Let's count how much customers buy our product

In [75]:
custumer_df.groupby(by="status").customer_id.count()

status
Active        617
Non Active    383
Name: customer_id, dtype: int64

To get much more information about the customer sales let's concat this two dataframe

In [82]:
customer_order_df = pd.merge(
    left=orders_df,
    right=custumer_df,
    how='left',
    left_on='customer_id',
    right_on='customer_id'
)
customer_order_df

order_id  customer_id  payment order_date delivery_date  delivery_time  \
0           1           64    30811 2021-08-30    2021-09-24           25.0   
1           2          473    50490 2021-02-03    2021-02-13           10.0   
2           3          774    46763 2021-10-08    2021-11-03           26.0   
3           4          433    39782 2021-05-06    2021-05-19           13.0   
4           5          441    14719 2021-03-23    2021-03-24            1.0   
..        ...          ...      ...        ...           ...            ...   
995       996          345    37843 2021-01-13    2021-02-02           20.0   
996       997          346    53831 2021-01-18    2021-01-31           13.0   
997       998          407    53308 2021-05-05    2021-05-21           16.0   
998       999          428    31643 2021-06-15    2021-07-12           27.0   
999      1000          896    27836 2021-04-07    2021-04-24           17.0   

    customer_name             gender  age                   home_address  \
0        fulan 64  Prefer not to say   75      4927 Alice MeadowApt. 960   
1       fulan 473               Male   61  531 Schmitt BoulevardApt. 010   
2       fulan 774  Prefer not to say   34       2096 Wilson MewsApt. 714   
3       fulan 433  Prefer not to say   46        5777 Mayer PassApt. 881   
4       fulan 441  Prefer not to say   53   33 Richards JunctionApt. 478   
..            ...                ...  ...                            ...   
995     fulan 345  Prefer not to say   62      9182 Hermann MallApt. 890   
996     fulan 346  Prefer not to say   40    59 Bergstrom GroveSuite 260   
997     fulan 407               Male   41      5233 Lara AvenueSuite 291   
998     fulan 428  Prefer not to say   66      80 Hermiston WaySuite 403   
999     fulan 896  Prefer not to say   67     985 Nguyen TerraceApt. 362   

     zip_code                city                         state    country  \
0        7787      Sanfordborough               South Australia  Australia   
1        1744             Annaton               South Australia  Australia   
2        8590  West Jonathanshire                      Tasmania  Australia   
3        9728    West Michaelport                      Tasmania  Australia   
4        7650     South Rileyview             Western Australia  Australia   
..        ...                 ...                           ...        ...   
995       799          Audreyfurt                      Victoria  Australia   
996      6060           Olsonbury            Northern Territory  Australia   
997      8697       Hartmannshire                    Queensland  Australia   
998      1112      North Liamfort  Australian Capital Territory  Australia   
999      7737            Ryanberg               South Australia  Australia   

     status  
0    Active  
1    Active  
2    Active  
3    Active  
4    Active  
..      ...  
995  Active  
996  Active  
997  Active  
998  Active  
999  Active  

[1000 rows x 15 columns]

Let's analyze this new dataframe with pivot table

#### Customer order by city

In [83]:
customer_order_df.groupby(by='city').order_id.nunique().sort_values(ascending=False)

city
Jordanside    6
New Ava       6
Lake Rose     5
O'keefeton    5
East Max      5
             ..
Noahmouth     1
Graceville    1
New Sophie    1
New Sienna    1
Zacville      1
Name: order_id, Length: 605, dtype: int64

#### Customer order by state

In [84]:
customer_order_df.groupby(by="state").order_id.nunique().sort_values(ascending=False)

state
South Australia                 148
Queensland                      139
Western Australia               130
New South Wales                 129
Australian Capital Territory    118
Victoria                        118
Tasmania                        112
Northern Territory              106
Name: order_id, dtype: int64

#### Customer order by age

In [85]:
customer_order_df.groupby(by="age").order_id.nunique().sort_values(ascending=False)

age
57    34
75    27
32    27
38    26
63    26
      ..
72     8
29     8
77     8
33     7
65     7
Name: order_id, Length: 61, dtype: int64

#### Customer order by gender

In [86]:
customer_order_df.groupby(by="gender").order_id.nunique().sort_values(ascending=False)

gender
Prefer not to say    725
Female               139
Male                 136
Name: order_id, dtype: int64

#### Customer order by age group

In [87]:
customer_order_df["age_group"] = customer_order_df.age.apply(lambda x: "Youth" if x <= 24 else ("Seniors" if x > 64 else "Adults"))
customer_order_df.groupby(by="age_group")["order_id"].nunique().sort_values(ascending=False)

age_group
Adults     681
Seniors    226
Youth       93
Name: order_id, dtype: int64

### Exploring Product Dataframe

This dataframe contains information of our product like size, colour, and price

In [89]:
product_df.describe(include='all')

product_id product_type product_name  size colour        price  \
count   1260.000000         1260         1260  1260   1260  1260.000000   
unique          NaN            3           35     8      7          NaN   
top             NaN        Shirt        Denim    XS    red          NaN   
freq            NaN          420           70   252    180          NaN   
mean     629.500000          NaN          NaN   NaN    NaN   105.805556   
std      363.874979          NaN          NaN   NaN    NaN     9.704423   
min        0.000000          NaN          NaN   NaN    NaN    90.000000   
25%      314.750000          NaN          NaN   NaN    NaN    95.750000   
50%      629.500000          NaN          NaN   NaN    NaN   108.500000   
75%      944.250000          NaN          NaN   NaN    NaN   114.000000   
max     1259.000000          NaN          NaN   NaN    NaN   119.000000   

           quantity                                   description  
count   1260.000000                                          1260  
unique          NaN                                          1260  
top             NaN  A red coloured, XS sized, Oxford Cloth Shirt  
freq            NaN                                             1  
mean      60.150000                                           NaN  
std       11.670573                                           NaN  
min       40.000000                                           NaN  
25%       50.000000                                           NaN  
50%       60.000000                                           NaN  
75%       70.000000                                           NaN  
max       80.000000                                           NaN

According to this table, we can know that our product price start from 90 to 119 usd with mean 105,80. 

Now we gonna search which product is the most expensive

In [92]:
product_df.sort_values(by='price', ascending=False)

product_id product_type product_name size  colour  price  quantity  \
704          698       Jacket        Parka    L  violet    119        53   
671          665       Jacket        Parka   XS     red    119        65   
698          692       Jacket        Parka    M  indigo    119        66   
699          693       Jacket        Parka    L  indigo    119        44   
700          694       Jacket        Parka   XL  indigo    119        78   
...          ...          ...          ...  ...     ...    ...       ...   
1225        1219     Trousers      Pleated   XL  indigo     90        45   
1226        1220     Trousers      Pleated   XS  violet     90        58   
1227        1221     Trousers      Pleated    S  violet     90        50   
1228        1222     Trousers      Pleated    M  violet     90        45   
599          593       Jacket       Bomber    L  violet     90        68   

                                        description  
704        A violet coloured, L sized, Parka Jacket  
671          A red coloured, XS sized, Parka Jacket  
698        A indigo coloured, M sized, Parka Jacket  
699        A indigo coloured, L sized, Parka Jacket  
700       A indigo coloured, XL sized, Parka Jacket  
...                                             ...  
1225  A indigo coloured, XL sized, Pleated Trousers  
1226  A violet coloured, XS sized, Pleated Trousers  
1227   A violet coloured, S sized, Pleated Trousers  
1228   A violet coloured, M sized, Pleated Trousers  
599       A violet coloured, L sized, Bomber Jacket  

[1260 rows x 8 columns]

Now we make a pivot table to revealed information about the product

In [93]:
product_df.groupby(by="product_type").agg({
    "product_id": "nunique",
    "quantity": "sum",
    "price": ["min", "max"]
})

product_id quantity price     
                nunique      sum   min  max
product_type                               
Jacket              420    25387    90  119
Shirt               420    25263    92  119
Trousers            420    25139    90  119

In [96]:
product_df.groupby(by="product_name").agg({
    "product_id": "nunique",
    "quantity": "sum",
    "price": ["min", "max"]
})

product_id quantity price     
                     nunique      sum   min  max
product_name                                    
Bomber                    35     2083    90   90
Camp Collared             35     2071   112  112
Cardigan                  35     2032   118  118
Cargo Pants               35     2146   106  106
Casual Slim Fit           35     2086   113  113
Chambray                  35     2020   105  105
Chinos                    35     2101   100  100
Coach                     35     2158   115  115
Cords                     35     2260   113  113
Cropped                   35     2085    99   99
Cuban Collar              35     2017    93   93
Denim                     70     4204    92  115
Drawstring                35     2042   104  104
Dress                     35     2125   119  119
Flannel                   35     2135    96   96
Henley                    35     2051    92   92
High-Waisted              35     2198    98   98
Joggers                   35     2107    94   94
Leather                   35     2276   113  113
Linen                     35     2138   116  116
Mandarin Collar           35     2203   108  108
Oxford Cloth              35     2071   114  114
Parka                     35     2201   119  119
Peacoat                   35     2067   102  102
Pleated                   35     2030    90   90
Polo                      35     2196   117  117
Puffer                    35     1946   110  110
Pullover                  35     2184   114  114
Relaxed Leg               35     2002    95   95
Shearling                 35     2169    95   95
Slim-Fit                  35     2155   119  119
Tracksuit Bottoms         35     2038    91   91
Trench Coat               35     2132   112  112
Windbreaker               35     2085   109  109
Wool                      35     1975   111  111

### Exploring Sales Dataframe

This dataframe contains information about the sales of the product, the buys quantity, and the price of the product that customer buy

In [94]:
sales_df.describe()

sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  5000.000000  
mean    206.360000  
std      86.357457  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

not much information we got from this dataframe, except we concat it with other dataframe

### Exploring Sales and Product Dataframe

In [97]:
sales_product_df = pd.merge(
    left=sales_df,
    right=product_df,
    how='left',
    left_on="product_id",
    right_on="product_id"
)
sales_product_df

sales_id  order_id  product_id  price_per_unit  quantity_x  total_price  \
0            0         1         218             106           2          212   
1            1         1         481             118           1          118   
2            2         1           2              96           3          288   
3            3         1        1002             106           2          212   
4            4         1         691             113           3          339   
...        ...       ...         ...             ...         ...          ...   
4995      4995       998         321             109           2          218   
4996      4996       998         251              95           3          285   
4997      4997       999         872             113           1          113   
4998      4998       999         998             106           2          212   
4999      4999       999        1105             115           1          115   

     product_type     product_name size  colour  price  quantity_y  \
0           Shirt         Chambray    L  orange    105          44   
1          Jacket           Puffer    S  indigo    110          62   
2           Shirt     Oxford Cloth    M     red    114          54   
3        Trousers             Wool    M    blue    111          52   
4          Jacket            Parka    S  indigo    119          53   
...           ...              ...  ...     ...    ...         ...   
4995        Shirt  Mandarin Collar    S  orange    108          52   
4996        Shirt             Polo    S  orange    117          78   
4997     Trousers           Chinos    M  violet    100          54   
4998     Trousers             Wool    L   green    111          42   
4999     Trousers          Cropped   XS    blue     99          47   

                                            description  
0            A orange coloured, L sized, Chambray Shirt  
1             A indigo coloured, S sized, Puffer Jacket  
2           A red coloured, M sized, Oxford Cloth Shirt  
3               A blue coloured, M sized, Wool Trousers  
4              A indigo coloured, S sized, Parka Jacket  
...                                                 ...  
4995  A orange coloured, S sized, Mandarin Collar Shirt  
4996             A orange coloured, S sized, Polo Shirt  
4997        A violet coloured, M sized, Chinos Trousers  
4998           A green coloured, L sized, Wool Trousers  
4999        A blue coloured, XS sized, Cropped Trousers  

[5000 rows x 13 columns]

Okay let's make a pivot table on this dataset

In [103]:
sales_product_df.groupby(by="product_type").agg({
    "sales_id": "nunique",
    "quantity_x": "sum",
    "total_price": "sum"
})

sales_id  quantity_x  total_price
product_type                                   
Jacket            1676        3343       357026
Shirt             1641        3259       333600
Trousers          1683        3360       341174

If we see in this pivot table we see that trousers is the most selling product, but Jacket gives more provit to corporate

In [106]:
sales_product_df.groupby(by="product_name").agg({
    "sales_id": "nunique",
    "quantity_x": "sum",
    "total_price": "sum"    
}).sort_values(by= "total_price", ascending=False)

sales_id  quantity_x  total_price
product_name                                        
Denim                   273         527        52399
Casual Slim Fit         154         306        36414
Trench Coat             146         299        35581
Shearling               150         302        35334
Puffer                  140         298        35164
Flannel                 141         281        33158
Cropped                 135         284        32660
Pleated                 147         308        32340
Joggers                 164         334        31062
Chambray                141         290        30740
Parka                   134         269        30397
Bomber                  150         297        29700
Chinos                  135         260        29380
Windbreaker             143         287        29274
Cardigan                134         260        29120
Peacoat                 132         266        28994
Tracksuit Bottoms       130         257        28784
Slim-Fit                154         288        27936
Oxford Cloth            146         289        27744
Drawstring              147         288        27648
High-Waisted            140         278        27522
Camp Collared           139         279        27342
Cargo Pants             134         267        27234
Relaxed Leg             146         301        27090
Cuban Collar            122         254        26670
Henley                  146         281        26414
Wool                    125         249        26394
Leather                 136         263        25774
Mandarin Collar         124         236        25724
Dress                   127         243        25272
Pullover                143         276        25116
Polo                    131         264        25080
Coach                   134         265        24645
Linen                   131         270        24570
Cords                   126         246        23124

We see that denim gives more provit to corporate than any other product

### Exploring All Dataframe

Okay now wee gonna merge all of our dataframe so that we know the demografi of our customer

In [107]:
all_df = pd.merge(
    left=sales_product_df,
    right=customer_order_df,
    how='left',
    left_on="order_id",
    right_on="order_id"
)

In [108]:
all_df

sales_id  order_id  product_id  price_per_unit  quantity_x  total_price  \
0            0         1         218             106           2          212   
1            1         1         481             118           1          118   
2            2         1           2              96           3          288   
3            3         1        1002             106           2          212   
4            4         1         691             113           3          339   
...        ...       ...         ...             ...         ...          ...   
4995      4995       998         321             109           2          218   
4996      4996       998         251              95           3          285   
4997      4997       999         872             113           1          113   
4998      4998       999         998             106           2          212   
4999      4999       999        1105             115           1          115   

     product_type     product_name size  colour  ...  customer_name  \
0           Shirt         Chambray    L  orange  ...       fulan 64   
1          Jacket           Puffer    S  indigo  ...       fulan 64   
2           Shirt     Oxford Cloth    M     red  ...       fulan 64   
3        Trousers             Wool    M    blue  ...       fulan 64   
4          Jacket            Parka    S  indigo  ...       fulan 64   
...           ...              ...  ...     ...  ...            ...   
4995        Shirt  Mandarin Collar    S  orange  ...      fulan 407   
4996        Shirt             Polo    S  orange  ...      fulan 407   
4997     Trousers           Chinos    M  violet  ...      fulan 428   
4998     Trousers             Wool    L   green  ...      fulan 428   
4999     Trousers          Cropped   XS    blue  ...      fulan 428   

                 gender age               home_address  zip_code  \
0     Prefer not to say  75  4927 Alice MeadowApt. 960      7787   
1     Prefer not to say  75  4927 Alice MeadowApt. 960      7787   
2     Prefer not to say  75  4927 Alice MeadowApt. 960      7787   
3     Prefer not to say  75  4927 Alice MeadowApt. 960      7787   
4     Prefer not to say  75  4927 Alice MeadowApt. 960      7787   
...                 ...  ..                        ...       ...   
4995               Male  41  5233 Lara AvenueSuite 291      8697   
4996               Male  41  5233 Lara AvenueSuite 291      8697   
4997  Prefer not to say  66  80 Hermiston WaySuite 403      1112   
4998  Prefer not to say  66  80 Hermiston WaySuite 403      1112   
4999  Prefer not to say  66  80 Hermiston WaySuite 403      1112   

                city                         state    country  status  \
0     Sanfordborough               South Australia  Australia  Active   
1     Sanfordborough               South Australia  Australia  Active   
2     Sanfordborough               South Australia  Australia  Active   
3     Sanfordborough               South Australia  Australia  Active   
4     Sanfordborough               South Australia  Australia  Active   
...              ...                           ...        ...     ...   
4995   Hartmannshire                    Queensland  Australia  Active   
4996   Hartmannshire                    Queensland  Australia  Active   
4997  North Liamfort  Australian Capital Territory  Australia  Active   
4998  North Liamfort  Australian Capital Territory  Australia  Active   
4999  North Liamfort  Australian Capital Territory  Australia  Active   

     age_group  
0      Seniors  
1      Seniors  
2      Seniors  
3      Seniors  
4      Seniors  
...        ...  
4995    Adults  
4996    Adults  
4997   Seniors  
4998   Seniors  
4999   Seniors  

[5000 rows x 28 columns]

Now let's create the pivot table

In [113]:
all_df.groupby(by=["state", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
state                        product_type                         
Australian Capital Territory Jacket               406        43204
                             Shirt                396        40448
                             Trousers             454        46790
New South Wales              Jacket               451        47998
                             Shirt                431        43980
                             Trousers             392        39766
Northern Territory           Jacket               365        38991
                             Shirt                336        33865
                             Trousers             384        38998
Queensland                   Jacket               499        53511
                             Shirt                417        42506
                             Trousers             453        46045
South Australia              Jacket               461        49090
                             Shirt                509        52685
                             Trousers             455        46041
Tasmania                     Jacket               412        44370
                             Shirt                387        39668
                             Trousers             379        38057
Victoria                     Jacket               359        38203
                             Shirt                390        40206
                             Trousers             414        41948
Western Australia            Jacket               390        41659
                             Shirt                393        40242
                             Trousers             429        43529

This table gives information about what's customer likes according to where their lives

Let's create more

In [114]:
all_df.groupby(by=["gender", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
gender            product_type                         
Female            Jacket               481        50963
                  Shirt                445        45146
                  Trousers             481        49295
Male              Jacket               480        50978
                  Shirt                410        41939
                  Trousers             453        46071
Prefer not to say Jacket              2382       255085
                  Shirt               2404       246515
                  Trousers            2426       245808

This table gives us information about what they like according to what their gender

In [115]:
all_df.groupby(by=["age_group", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
age_group product_type                         
Adults    Jacket              2292       245055
          Shirt               2225       227781
          Trousers            2272       231271
Seniors   Jacket               777        82959
          Shirt                733        74977
          Trousers             766        77471
Youth     Jacket               274        29012
          Shirt                301        30842
          Trousers             322        32432

This table gives us information about what they likes according to what group of age they are